### PR TITLE
feat(ingest): add more fail-safes to stateful ingestion

### PR DESCRIPTION
**Pull request by hsheth2**
Stacked on top of the changes in #8104. Review that PR first.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py [hunk 1]
```diff
@@ -217,11 +218,30 @@ def gen_removed_entity_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+            self.source.get_report().report_failure(
+                "stale-entity-removal",
+                "Skipping stale entity soft-deletion because the source produced no events. "
+                "This is a fail-safe mechanism to prevent accidental deletion of all entities.",
+            )
+            return
+
+        # If the source already had a failure, skip soft-deletion.
+        # TODO: Eventually, switch this to check if anything in the pipeline had a failure so far.
+        if self.source.get_report().failures:
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py:240]: Does this take into account the `commit_policy`? Or for stale entity removal the rule is just commit iff there are no failures

**Review by asikowitz — APPROVED** [review]
One question but lgtm

File: metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py [hunk 2]
```diff
@@ -191,22 +191,23 @@ def gen_removed_entity_workunits(self) -> Iterable[MetadataWorkUnit]:
...
             self.job_id, self.state_type_class
         )
         if not last_checkpoint:
             return
         cur_checkpoint = self.state_provider.get_current_checkpoint(self.job_id)
         assert cur_checkpoint is not None
         # Get the underlying states
-        last_checkpoint_state = cast(GenericCheckpointState, last_checkpoint.state)
+        last_checkpoint_state: GenericCheckpointState = last_checkpoint.state
         cur_checkpoint_state = cast(GenericCheckpointState, cur_checkpoint.state)
```
[the remark this answers is not in the document]
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py:220]: nope - get_current_checkpoint doesn't infer types in the same way

File: metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py (hunk 1 above)
Comment (asikowitz) [reply metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py:257]: Does that mean we're dropping support for a configurable commit policy? Right now it seems like it doesn't really do anything

Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py:257]: eventually yes I'd like to drop support for that

**Pull request merged**